### PR TITLE
RSDK-8306 - set version in metadata

### PR DESCRIPTION
**Pull request by stuqdog**

### CODE REVIEW

File: src/viam/sdk/tests/mocks/mock_robot.cpp [hunk 1]
```diff
@@ -265,9 +269,14 @@ std::vector<FrameSystemConfig> mock_proto_config_response() {
...
 ::grpc::Status MockRobotService::FrameSystemConfig(
-    ::grpc::ServerContext*,
+    ::grpc::ServerContext* context,
     const ::viam::robot::v1::FrameSystemConfigRequest*,
     ::viam::robot::v1::FrameSystemConfigResponse* response) {
+    auto client_info = context->client_metadata().find("viam_client");
+    if (client_info->second != k_version) {
+        return ::grpc::Status(::grpc::StatusCode::FAILED_PRECONDITION,
+                              "viam_client info not properly set in metadata");
+    }
```
Comment (stuqdog) [inline src/viam/sdk/tests/mocks/mock_robot.cpp:279]: These robot tests gave the easiest access to server-side context such that we can confirm that the `viam_client` header is properly being set.

File: src/viam/sdk/tests/test_utils.cpp [hunk 2]
```diff
@@ -61,7 +61,7 @@ std::vector<GeometryConfig> fake_geometries() {
             std::move(capsule_config)};
 }
 
-TestServer::TestServer(std::shared_ptr<Server> sdk_server) : sdk_server_(sdk_server) {}
+TestServer::TestServer(std::shared_ptr<Server> sdk_server) : sdk_server_(std::move(sdk_server)) {}
```
Comment (stuqdog) [inline src/viam/sdk/tests/test_utils.cpp:64]: flyby

File: .github/workflows/update_protos.yml [hunk 3]
```diff
@@ -46,12 +46,14 @@ jobs:
...
         run: rm -rf boost/ && rm -rf vcpkg/
+      - name: Update proto tag
+        run: sed -i -e 's/VIAMCPPSDK_API_PROTO_LABEL "v[0-9]*\.[0-9]*\.[0-9]*"/VIAMCPPSDK_API_PROTO_LABEL "${{ github.event.client_payload.tag }}"/' src/viam/sdk/CMakeLists.txt
       - name: Add + Commit + Open PR
         uses: peter-evans/create-pull-request@v3
         with:
           commit-message: '[WORKFLOW] Updating protos from ${{ github.event.client_payload.repo_name }}, commit: ${{ github.event.client_payload.sha }}'
           branch: 'workflow/update-protos'
-          add-paths: src/viam/api/*
+          add-paths: src/viam/api/*, src/viam/sdk/CMakeLists.txt
```
Comment (stuqdog) [inline .github/workflows/update_protos.yml:56]: Unfortunately this is only partially untested. I know that the `sed` command works locally and I know that `tag` is the correct payload name based on the `api` job, but we're currently already caught up in proto here. I have fairly high confidence that this will work and I believe the consequences of being wrong are low (next time `update_protos` runs we'd either see it fail or improperly update the `VIAMCPPSDK_API_PROTO_LABEL` variable, either of which should be fairly straightforward to fix), but if anyone has concerns let me know!

File: src/viam/sdk/common/version.cpp.in [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#include <viam/sdk/common/version.hpp>
+
+namespace viam {
+namespace sdk {
+
+const char* const k_version = "cpp;@VIAMCPPSDK_SDK_VERSION@;@VIAMCPPSDK_API_PROTO_LABEL@";
```
Comment (acmorrow) [inline src/viam/sdk/common/version.cpp.in:6]: - I don't think this string belongs as part of the public interface of the SDK, since it is entirely an implementation detail. Consider creating a `viam/sdk/common/private/version_metadata.[hpp,cpp]` instead.
- I don't think you need the indirection through a pointer: `constexpr char k_version_metadata[] = "..."`.
- We probably someday will want a public `version.hpp` which contains the major/minor/patch versions broken out, along with some convenience methods.

File: src/viam/api/CMakeLists.txt [hunk 5]
```diff
@@ -261,7 +261,7 @@ endif(VIAMCPPSDK_USE_DYNAMIC_PROTOS)
 
 add_custom_target(
   update-buf
-  COMMAND buf mod update ${CMAKE_CURRENT_SOURCE_DIR}
+  COMMAND buf dep update ${CMAKE_CURRENT_SOURCE_DIR}
```
[the remark this answers is not in the document]
Comment (stuqdog) [reply src/viam/api/CMakeLists.txt:264]: Got it! I didn't realize there was complication here, happy to revert this change.

File: src/viam/sdk/CMakeLists.txt [hunk 6]
```diff
@@ -19,6 +19,11 @@ target_include_directories(viamsdk
     "$<BUILD_INTERFACE:$<TARGET_PROPERTY:viam-cpp-sdk::viamapi,INTERFACE_INCLUDE_DIRECTORIES>>"
 )
 
+# Configure the SDK version file
+set(VIAMCPPSDK_API_PROTO_LABEL "v0.1.327")
+set(VIAMCPPSDK_SDK_VERSION ${CMAKE_PROJECT_VERSION})
```
[the remark this answers is not in the document]
Comment (stuqdog) [reply src/viam/sdk/CMakeLists.txt:24]: We don't! It is more explicit but also more verbose, can definitely just use `CMAKE_PROJECT_VERSION` for both.

File: src/viam/sdk/common/version.cpp.in (hunk 4 above)
Comment (stuqdog) [reply src/viam/sdk/common/version.cpp.in:6]: Cool, will make this private, remove the pointer-ness of it, and create a ticket for an eventual public `version.hpp`!

File: .github/workflows/update_protos.yml [hunk 7]
```diff
@@ -46,12 +46,14 @@ jobs:
           BOOST_ROOT: ${{ github.workspace }}/boost/boost
       - name: cleanup
         run: rm -rf boost/ && rm -rf vcpkg/
+      - name: Update proto tag
```
[the remark this answers is not in the document]
Comment (stuqdog) [reply .github/workflows/update_protos.yml:49]: Yeah so I had a little bash script for getting the version tag based on the commit in the `buf.lock` file but I ended up going against that approach because it felt a little fragile (required internet access, relied on the output of `buf beta registry label list` maintaining a consistent format). I didn't see any other good way to get the tag (it doesn't seem to exist in the sources anywhere).

> Also, how does it work above that doing the `build --target update-static-protos` is building against the version of the protos that we are then interpolating?

Sorry, I'm not sure I understand what you're asking here exactly. Can you rephrase?

File: src/viam/api/CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (stuqdog) [reply src/viam/api/CMakeLists.txt:264]: Eh yeah, that doesn't sound great. I had assumed that the swap to `dep` would be harmless but given that it's not and that the warning explicitly says that `buf mod update` will continue to work, I think it's fine to just stick with what we had before for now.

File: src/viam/sdk/common/version.cpp.in (hunk 4 above)
Comment (stuqdog) [reply src/viam/sdk/common/version.cpp.in]: ticket: https://viam.atlassian.net/browse/RSDK-8343

Comment (lia-viam) [reply src/viam/sdk/common/version.cpp.in]: If we're making it private we might consider whether we want to skip the header/source split entirely, see eg 
https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/sdk/common/utils.hpp#L20
making it `const` gives it the right linkage anyway, and if it's only ever included in a `.cpp` file I don't think we gain anything from breaking it off compilation-wise. And then I think it's a tossup whether we want it `std::string` or `char` ptr/array

Comment (acmorrow) [reply src/viam/sdk/common/version.cpp.in]: I wish we had `string_view` because a `constexpr string_view` is really the best thing for this. I don't care too much if we want to make it a std::string, though I don't think it buys us much over a char array in this particular case. I agree that, as a private header, we can certainly avoid the need for a separate .cpp. I would like to see the file renamed from `version.hpp` to `version_metadata.hpp` though.

File: .github/workflows/update_protos.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (acmorrow) [reply .github/workflows/update_protos.yml:49]: > > Also, how does it work above that doing the `build --target update-static-protos` is building against the version of the protos that we are then interpolating?
> 
> Sorry, I'm not sure I understand what you're asking here exactly. Can you rephrase?

Let me try again. Basically, I'd like to understand better what the workflow does. How does it end up building the protos against a specific version of the API repo? Can *that* process be adjusted such that the API repo version tag can be programatically read out of the static protos in the tree, rather than needing to edit a constant in the CMakeLists.

File: src/viam/sdk/common/version.cpp.in (hunk 4 above)
Comment (lia-viam) [reply src/viam/sdk/common/version.cpp.in]: There is `boost::core::string_view` although it may be a bit silly to use just for this thing

Comment (acmorrow) [reply src/viam/sdk/common/version.cpp.in]: Yes, but if we want it, I'd like to either 1) go to C++17 as the language standard of record for the project and then just use it, or 2) polyfill it in to use either C++17 when built that way or boost when built for C++14. So, let's just defer that and stay with either `std::string` or `char` arrray for now.

Comment (stuqdog) [reply src/viam/sdk/common/version.cpp.in]: renamed to `version_metadata.hpp`, sticking with `char` now for simplicity. Do we want to get rid of the `.cpp` file entirely? It makes sense to me but I don't have a particular opinion either way, happy to defer here.

File: .github/workflows/update_protos.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (stuqdog) [reply .github/workflows/update_protos.yml:49]: Ah, the workflow builds the latest version and runs whenever the protos are updated, then updates the commit in `buf.lock` to be whatever the latest commit is. It always builds against main. It gets the tag info from the `api` repo's `tag_and_release_protos` workflow. 

I'm not aware of anywhere currently where the tag exists in the tree such that we could get it without storing a constant somewhere. We _could_ write it to a separate file, but then we're still editing a constant somewhere and it complicates the process of building out the `version_metadata.cpp` file since we'd then be modifying it twice from two different sources and have to preserve both changes, so we'd probably need to have an intermediary file in between `version_metadata.cpp.in` and `version_metadata.cpp`. So I'm not sure if that's a great approach.

Alternatively, we could use the `buf` command line tool to get tag info and then some grep/sed magic to get the tag, but again I think this is pretty unreliable and probably not worth it.

File: src/viam/sdk/common/utils.cpp [hunk 8]
```diff
@@ -15,6 +15,7 @@
 
 #include <viam/api/common/v1/common.pb.h>
 
+#include <viam/sdk/common/version.hpp>
```
[the remark this answers is not in the document]
Comment (stuqdog) [reply src/viam/sdk/common/utils.cpp]: Yep! quite right, fixing now :)

File: src/viam/sdk/common/private/version.cpp.in [hunk 9]
```diff
@@ -0,0 +1,9 @@
+#include <viam/sdk/common/private/version.hpp>
```
[the remark this answers is not in the document]
Comment (stuqdog) [reply src/viam/sdk/common/private/version.cpp.in]: So I've tried moving all this into the header and changing it to `version_metadata.hpp.in` but I'm getting a compiler error `viam/sdk/common/private/version_metadata.hpp file not found`. It obviously doesn't exist in the main tree (`version_metadata.hpp.*in*` does), but the file at that path _does_ exist in the `build` dir. Feel like I'm doing something obviously wrong, but I'm not sure what.

File: src/viam/sdk/CMakeLists.txt [hunk 10]
```diff
@@ -17,8 +17,13 @@ add_library(viam-cpp-sdk::viamsdk ALIAS viamsdk)
 target_include_directories(viamsdk
   PUBLIC
     "$<BUILD_INTERFACE:$<TARGET_PROPERTY:viam-cpp-sdk::viamapi,INTERFACE_INCLUDE_DIRECTORIES>>"
+  PRIVATE "${CMAKE_CURRENT_BINARY_DIR}/../.."
```
Comment (stuqdog) [inline src/viam/sdk/CMakeLists.txt:20]: This makes it so the compiler can see the generated `version_metadata.hpp` file when building. It intuitively feels wrong to me to include the build dir here (even privately), but this is the solution proposed by the [cmake documentation](https://cmake.org/cmake/help/latest/command/configure_file.html#example), so :shrug:. My discomfort is naive (I don't really know well enough to say concretely _why_ I don't like it other than it just gives off a bit of a code smell), so if anyone has thoughts or suggestions I'm all ears.

The `/../..` suffix is necessary to get the include to look right/consistent in `utils.hpp`, otherwise the include path would be inconsistent with every other SDK file.

File: src/viam/sdk/tests/mocks/mock_robot.cpp [hunk 11]
```diff
@@ -1,9 +1,13 @@
...
 #include <viam/sdk/tests/mocks/mock_robot.hpp>
 
+#include <vector>
+
+#include <grpcpp/support/status.h>
+
 #include <common/v1/common.pb.h>
 #include <robot/v1/robot.pb.h>
 
+#include <src/viam/sdk/common/private/version_metadata.hpp>
```
Comment (stuqdog) [inline src/viam/sdk/tests/mocks/mock_robot.cpp:10]: I cannot figure out why this has to be `#include <*src*/viam/sdk...>` when every other SDK include is `#include <viam/sdk/...>`. I tried messing around with `include_directories` in the test `CmakeLists.txt`, but couldn't figure this out and struggled to find any guidance online. If anyone has thoughts here I'd appreciate them!

File: .github/workflows/update_protos.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (acmorrow) [reply .github/workflows/update_protos.yml:49]: @stuqdog and I talked this over. Writing the version to a file under `src/viam/api` will simplify the script and also not give the appearance that this value is actually "always" true. It isn't true, for instance, if you are building against local protos, in which case the API version should be "unknown" or something more like `git describe` output.

File: src/viam/sdk/CMakeLists.txt (hunk 10 above)
Comment (acmorrow) [reply src/viam/sdk/CMakeLists.txt:20]: I think this is expected  for a generated file, but I will take a quick look to see if there is a more idiomatic way to express it.

File: src/viam/sdk/tests/CMakeLists.txt [hunk 12]
```diff
@@ -17,6 +17,10 @@ enable_testing()
 include(viamcppsdk_utils)
 add_library(viamsdk_test)
 
+target_include_directories(viamsdk_test
+  PRIVATE "${CMAKE_CURRENT_BINARY_DIR}/../../.."
+)
+
```
[the remark this answers is not in the document]
Comment (acmorrow) [reply src/viam/sdk/tests/CMakeLists.txt:23]: Discussed this with @stuqdog: given that there are situations where the value may not match the one pulled from the file (e.g. you built against local protos), the tests will just ensure that an API version of some sort is present but not compare it.

File: .github/workflows/update_protos.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (njooma) [reply .github/workflows/update_protos.yml:49]: Do we want to always build against main? I feel like we should only build against the version tag that was provided. I think every time someone builds a proto, it should be the same, predictable output regardless of what's going on in API world. Only when the tag is changed (either manually or via this workflow) should the compiled output change.

[the remark this answers is not in the document]
Comment (stuqdog) [reply .github/workflows/update_protos.yml:49]: In general, every time someone builds a proto it _is_ the same, predictable output. The one exception that we're worried about here is when someone explicitly opts in to building against a local API definition (if, e.g., they wanted to test out API changes within the C++ SDK). This is explicitly opt-in and unless someone is specifically looking for it they'll never encounter or have to think about it. If we no longer want to offer that as an option we can certainly discuss, but I'd argue that such a change is out of scope for this PR.

File: src/viam/sdk/CMakeLists.txt (hunk 10 above)
Comment (acmorrow) [reply src/viam/sdk/CMakeLists.txt:20]: I pulled your branch to experiment. My findings are that 1) yes this is needed, but 2) there is a somewhat better way to write it, and 3) it should be moved elsewhere in this CMakeLists.txt file.

This particular call to `target_include_directories` is doing something tricky to ensure that the proto paths come first, for reasons that are outlined in the comment above.

Instead, you can add a variant of this line to the *other* `target_include_directories` call in this file, down around line 194 (new), and you can write it like this:

```
target_include_directories(...
    ...
    PRIVATE
        "$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/../..>"
```

Now, one thing I'm trying to figure out is how we manage to already have what is effectively `"$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../..>"` in our include directories, which is what makes the build work at all against the non-generated sources. But it is clearly necessary to tell cmake that we want to look in our build directory for this generated file, and this is the correct way to do it (scoped private, using a build-time generator).

Comment (stuqdog) [reply src/viam/sdk/CMakeLists.txt]: Awesome thanks for digging! Just updated.

**Review by acmorrow — APPROVED** [review]
LGTM

**Comment by stuqdog** [comment]
Screenshot 2024-07-29 at 12 00 18
screenshot of metadata being read in RDK, cc @njooma
